
# Geometri

**Læringsmål:**

* Forstå at kode må være feilfri for å kjøre, og rette enkle syntaksfeil

* Debugging

I denne oppgaven skal vi rette feil i koden og endre utskriftsformat.

Norsklæreren din kunne utmerket godt lest en stil og forstått den på tross av noen småfeil, og til og med gitt god karakter hvis innholdet forøvrig var bra. Datamaskinen er ikke like tilgivende: I et Python-program må alt være pinlig nøyaktig, en enkelt skrivefeil kan være nok til at programmet slett ikke kjører.

## a)

Rett alle syntaksfeilene i koden under slik at koden kjører uten problemer. Hver av syntaksfeilene her kan rettes ved å fjerne, legge til eller endre ett enkelt tegn, dvs. det er ikke nødvendig å gjøre større endringer i koden. Ved å kjøre programmet (velg ”Run” i menyen) og se på feilmeldinger vil du få hint om hvor hver enkelt syntaksfeil befinner seg.

In [11]:
r = 5
print("Vi har en sirkel med radius",  str(r))
omkrets = 2 * 3.14 * r
formatted_omkrets = "{:.2f}".format(omkrets)
print("Omkretsen er", formatted_omkrets)
areal = 3.14 * r ** 2
print("Arealet er", areal)
høyde = 8
volum = areal * høyde
print("Sylinder med høyde", str(høyde) + ":Volumet er",volum)

Vi har en sirkel med radius 5
Omkretsen er 31.40
Arealet er 78.5
Sylinder med høyde 8:Volumet er 628.0


Når koden fungerer bør du få følgende output:

  
```python
Vi har en sirkel med radius 5
Omkretsen er 31.400000000000002
Arealet er 78.5
Sylinder med høyde 8 : Volumet er 628.0
    ```

#### Hint

Tenk på parenteser, fnutter, variabelnavn og (+/,) i kombinasjon av tekst og variabelnavn.

#### Formatering av antall desimaler

At omkretsen får en liten hale med ...002 skyldes at flyttall ikke kan lagres helt nøyaktig i dataens minne. Derfor kan det være ønskelig med kun to desimaler. Det er en innebygd funksjon i Python som tar seg av formatering og den ser slik ut: `format(tall, formatering)`. `tall` er her tallet som skal konverteres, dvs. omkretsen i eksempelet over, mens formatering går ut på hvordan man ønsker å formere tallet. Et eksempel er format`(2/3, ‘.3f’)` som gir `0.667`. 3-tallet indikerer at det ønskes 3 desimaler.

## b

Endre koden fra **a)** til å skrive ut omkretsen med kun 2 desimaler. Dvs. at linje to over skal bli endret til Omkretsen er 31.40.

Kjøring av oppdatert kode:

  
```python
Vi har en sirkel med radius 5
Omkretsen er 31.40
Arealet er 78.5
Sylinder med høyde 8 : Volumet er 628.0
    ```

## c)

(vanskelig oppgave) I oppgave a får omkretsen verdien 31.400...002, hvorfor blir ikke omkretsen sin verdi lik 31.400...000?

Dobbelklikk på teksten under og skriv svaret ditt der:

> Skriv svaret ditt her
Dette skyldes begrensninger i representasjonen av flyttall i datamaskinen. Flyttall i de fleste programmeringsspråk, inkludert Python, bruker en representasjon kalt "floating-point" (flyttall) som bruker en begrenset mengde binære sifre for å representere et desimaltall.

I tilfellet med omkretsen av en sirkel, hvor vi beregner omkretsen som 2 * 3.14 * r, bruker vi en tilnærming av pi (π) med 3.14. Selv om pi er et irrasjonelt tall med uendelig mange desimaler, er det umulig å representere det nøyaktig med et endelig antall binære sifre.

Siden representasjonen av flyttall har en begrenset nøyaktighet, vil det alltid være en liten avrundingsfeil i flyttallsberegninger. Dette kan føre til at vi får et resultat som ikke er nøyaktig 31.4 når vi beregner omkretsen. I stedet kan det være en verdi som 31.400...002 eller 31.400...001.

For å håndtere dette, kan vi formatere resultatet til et bestemt antall desimaler når vi skriver det ut, som vi gjorde i tidligere svar ved hjelp av round() eller .format()-metoden. Dette lar oss begrense antall desimaler og få en mer ønsket representasjon av tallet, for eksempel 31.4.